statev2/src/replication/log_store.rs

```diff
@@ -216,18 +220,315 @@ impl Storage for LogStore {
+    /// i.e. upon raft initial startup
+    #[test]
+    fn test_startup_state() {
+        let store = mock_log_store();
+        let state = store.initial_state().unwrap();
+
+        assert_eq!(state.hard_state, HardState::new());
+        assert_eq!(state.conf_state, ConfState::new());
+    }
+
+    /// Tests applying a snapshot then fetching initial state, simulating a crash recovery
+    #[test]
+    fn test_recover_snapshot_state() {
+        let store = mock_log_store();
+        let snap = mock_snapshot();
+        apply_snapshot(&store, snap.clone());
+
+        // Now fetch the initial state
+        let state = store.initial_state().unwrap();
+
+        assert_eq!(state.hard_state.term, snap.get_metadata().get_term());
+        assert_eq!(state.hard_state.commit, snap.get_metadata().get_index());
+        assert_eq!(&state.conf_state, snap.get_metadata().get_conf_state());
+    }
+
+    /// Tests fetching a snapshot from storage when the snapshot is not stored
+    #[test]
+    fn test_missing_snapshot() {
+        let store = mock_log_store();
+        let res = store.snapshot(10 /* index */, 0 /* peer_id */);
+
+        assert!(matches!(
+            res,
+            Err(RaftError::Store(
+                RaftStorageError::SnapshotTemporarilyUnavailable
+            ))
+        ))
+    }
+
+    /// Tests fetching a snapshot from storage when the stored snapshot is out of date
+    #[test]
+    fn test_out_of_date_snapshot() {
+        let store = mock_log_store();
+        let snap = mock_snapshot();
+        apply_snapshot(&store, snap.clone());
+
+        // Attempt to fetch a snapshot at a higher index than the one stored
+        let index = snap.get_metadata().get_index() + 1;
+        let res = store.snapshot(index, 0 /* peer_id */);
+
+        assert!(matches!(
+            res,
+            Err(RaftError::Store(RaftStorageError::SnapshotOutOfDate))
+        ))
+    }
+
+    /// Tests fetching an up-to-date snapshot
+    #[test]
+    fn test_up_to_date_snapshot() {
+        let store = mock_log_store();
+        let snap = mock_snapshot();
+        apply_snapshot(&store, snap.clone());
+
+        // Attempt to fetch a snapshot at a lower index than the one stored
+        let index = snap.get_metadata().get_index() - 1;
+        let res = store.snapshot(index, 0 /* peer_id */);
+
+        assert!(res.is_ok());
+        let snap_res = res.unwrap();
+
+        assert_eq!(snap_res.get_metadata(), snap.get_metadata());
+    }
+
+    // -------------------
+    // | Log Entry Tests |
+    // -------------------
+
+    /// Tests fetching the first and last entries from an empty log
+    #[test]
+    fn test_empty_log() {
+        let store = mock_log_store();
+
+        let first = store.first_index().unwrap();
+        let last = store.last_index().unwrap();
+        let entry_term = store.term(1 /* index */);
+
+        assert_eq!(first, 0);
+        assert_eq!(last, 0);
```

Comment:
curious as to why the desired behavior is return `0` when a log at the given index doesn't exist, why don't we return a `RaftStorageError::Unavailable`?

Reply:
The consumer of this interface is the raft consensus engine, if we return error that signals a fault to the engine whereas if we return `0`, the engine will know the log to be empty